crates/full-node/sov-sequencer/src/preferred/rate_limiter/limiter.rs

```diff
@@ -0,0 +1,465 @@
+    pub(crate) drain_rate: DrainRatePerMillis<S::Gas>,
+}
+
+/// The amount of resources used by a request that needs to be rate-limited.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub(crate) struct ResourceUsed<G: Gas> {
+    pub(crate) inner: Resource<G>,
+}
+
+/// Defines the [`DrainRatePerMillis`] parameter for the token bucket algorithm. The [`RateLimiter`]
+/// tracks the total amount of resources consumed by each key (such as an IP address) and determines
+/// whether the resource usage exceeds a predefined threshold.
+///
+/// Meanwhile, the resources used by a given key are also continuously reduced at the rate specified by
+/// [`DrainRatePerMillis`] every millisecond.
+#[derive(Debug, Clone, Copy)]
+pub(crate) struct DrainRatePerMillis<G: Gas> {
+    pub(crate) resource_per_ms: Resource<G>,
+}
+
+impl<G: Gas> DrainRatePerMillis<G> {
+    /// We decrease the resource usage for each [`RateLimiter`] entry every millisecond.
+    fn mul_by_millis(&self, since_last_drain: u64) -> TotalResources<G> {
+        TotalResources {
+            inner: self
+                .resource_per_ms
+                .saturating_mul_by_scalar(since_last_drain),
+        }
+    }
+}
+
+/// The total resource usage for a given key (see [`Throttler`] bellow]).
+#[derive(Debug, Clone, Copy)]
+pub(crate) struct TotalResources<G: Gas> {
+    pub(crate) inner: Resource<G>,
+}
+
+impl<G: Gas> TotalResources<G> {
+    fn zero() -> Self {
+        Self {
+            inner: Resource::zero(),
+        }
+    }
+
+    fn combine(&mut self, used: &ResourceUsed<G>) {
+        self.inner = self.inner.checked_add(&used.inner).unwrap();
+    }
+
+    fn drain(&mut self, other: &Self) {
+        self.inner = self.inner.saturating_sub(&other.inner);
+    }
+
+    fn allow(&self, max: &Self) -> Result<(), LimitExceeded<G>> {
+        self.inner.err_if_exceeding(&max.inner)
+    }
+}
+
+/// Tracks the resources consumed for a given key, along with the last time total_resource_used was drained.
+#[derive(Debug, Clone, Copy)]
+pub(crate) struct Throttler<G: Gas> {
+    total_resource_used: TotalResources<G>,
+    last_drain: Instant,
+}
+
+impl<G: Gas> Throttler<G> {
+    /// Checks whether a new request can be processed and frees resources for the throttler
+    /// proportional to the amount of time that has passed since the last drain.
+    fn allow_request_and_drain_resource_used(
+        &mut self,
+        now: Instant,
+        max_allowed: &TotalResources<G>,
+        drain_rate: &DrainRatePerMillis<G>,
+    ) -> Result<(), LimitExceeded<G>> {
+        self.drain_total_resource_used(now, drain_rate);
+        self.total_resource_used.allow(max_allowed)
+    }
+
+    /// Increase the resource used by a given trottler.
+    fn throttle(&mut self, resource_used: ResourceUsed<G>) {
+        self.total_resource_used.combine(&resource_used);
+    }
+
+    fn drain_total_resource_used(&mut self, now: Instant, drain_rate: &DrainRatePerMillis<G>) {
```

Comment:
nit: terminology for token bucket is usually "refilling" when you're crediting people for not using resources and `draining` when they consume resources

Reply:
Oh, maybe the terminology is correct but the usage is backwards? It looks like we're draining based on how long it's been since the last time someone consumed resources. (`mul_by_millis(since_last_refill)`). Shouldn't we be first crediting their account with the refil accumulated in that time, then draining a fixed amount for their new consumption?